lib/Sema/BuilderTransform.cpp

```diff
@@ -296,99 +392,132 @@ class BuilderClosureVisitor
+        payloadIndex, numPayloads, isOptional);
 
     // Prepare the `else operand:
     Expr *elseExpr;
     SourceLoc elseLoc;
 
     // - If there's no `else` clause, use `Optional.none`.
-    if (!elseChain) {
+    if (!elseChainVar) {
       assert(isOptional);
       elseLoc = ifStmt->getEndLoc();
       elseExpr = buildNoneExpr(elseLoc);
 
     // - If there's an `else if`, the chain expression from that
     //   should already be producing a chain result.
     } else if (isElseIf) {
-      elseExpr = *elseChain;
+      elseExpr = buildVarRef(*elseChainVar, ifStmt->getEndLoc());
       elseLoc = ifStmt->getElseLoc();
 
     // - Otherwise, wrap it to produce a chain result.
     } else {
       elseLoc = ifStmt->getElseLoc();
-      elseExpr = buildWrappedChainPayload(*elseChain,
-                                          payloadIndex + 1, numPayloads,
-                                          isOptional);
+      elseExpr = buildWrappedChainPayload(
+          buildVarRef(*elseChainVar, ifStmt->getEndLoc()),
+          payloadIndex + 1, numPayloads, isOptional);
+    }
+
+    // Generate constraints for the various subexpressions.
+    auto condExpr = getTrivialBooleanCondition(ifStmt->getCond());
+    assert(condExpr && "Cannot get here without a trivial Boolean condition");
+    condExpr = cs->generateConstraints(condExpr, dc);
+    if (!condExpr) {
+      hadError = true;
+      return nullptr;
+    }
+
+    // Condition must convert to Bool.
+    // FIXME: This should be folded into constraint generation for conditions.
+    auto boolDecl = ctx.getBoolDecl();
+    if (!boolDecl) {
+      hadError = true;
+      return nullptr;
+    }
+    cs->addConstraint(ConstraintKind::Conversion,
+                      cs->getType(condExpr),
+                      boolDecl->getDeclaredType(),
+                      cs->getConstraintLocator(condExpr));
+
+    // The operand should have optional type if we had optional results,
+    // so we just need to call `buildIf` now, since we're at the top level.
+    if (isOptional && isTopLevel) {
+      thenExpr = buildCallIfWanted(ifStmt->getEndLoc(), ctx.Id_buildIf,
+                                   thenExpr,  /*argLabels=*/{ });
+      elseExpr = buildCallIfWanted(ifStmt->getEndLoc(), ctx.Id_buildIf,
+                                   elseExpr,  /*argLabels=*/{ });
+    }
+
+    thenExpr = cs->generateConstraints(thenExpr, dc);
+    if (!thenExpr) {
+      hadError = true;
+      return nullptr;
     }
 
-    Expr *condition = getTrivialBooleanCondition(ifStmt->getCond());
-    assert(condition && "checked by isBuildableIfChain");
+    elseExpr = cs->generateConstraints(elseExpr, dc);
+    if (!elseExpr) {
+      hadError = true;
+      return nullptr;
+    }
+
+    // FIXME: Need a locator for the "if" statement.
+    Type resultType = cs->addJoinConstraint(nullptr,
+        {
+          { cs->getType(thenExpr), cs->getConstraintLocator(thenExpr) },
```

Comment:
using `TernaryBranch(bool)` locator element would help to get better diagnostics in there is a mismatch here just like in `ConstraintGenerator::visitIfExpr`.

Reply:
Okay, thank you! I'll do that in a follow-up PR.

Reply:
Bah, I still need an "anchor", which pretty much has to be the "if" statement for this to make sense. I'll leave it as-is for the moment.